src/jit/lowerxarch.cpp

```diff
@@ -1575,6 +1575,15 @@ void Lowering::TreeNodeInfoInitBlockStore(GenTreeBlk* blkNode)
                     blkNode->gtLsraInfo.setInternalCandidates(l, l->internalFloatRegCandidates());
                 }
                 blkNode->gtBlkOpKind = GenTreeBlk::BlkOpKindUnroll;
+
+#ifdef _TARGET_X86_
+                // On x86, you can't address the lower byte of ESI, EDI, ESP, or EBP when doing
+                // a "mov byte ptr [dest], val". If the fill size is odd, we will try to do this
+                // when unrolling, so only allow byteable registers as the source value. (We could
+                // consider just using BlkOpKindRepInstr instead.)
+                sourceRegMask = RBM_BYTE_REGS;
```

Comment:
Shoot. In my head I'd written it to only do this for odd sizes, but apparently that thinking didn't make it into the file. I'll fix.